### PR TITLE
Implement migrator for old approval data

### PR DESCRIPTION
**Pull request by Tomin1**
This implements migrator component that reads through approval data from
sailjail homescreen plugin and stores approved permissions and
applications. After application's data has been migrated and saved the
old data is purged. Migration takes place on service start and happens
asynchronously one application at a time.

Migrator has state machine based on the one in prompter. There are separate
approval file queues for migrating and removing already migrated files. Once
a file has been migrated, it's put to removal queue so that once settings are
saved, it will be removed. As the migrations should happen pretty fast on
startup, in practice all the migration data is removed on one go in migrated
state. State machine and queues try to ensure that this will not block main
loop for too long.

### CODE REVIEW

File: daemon/migrator.c
Comment (Tomin1) [inline daemon/migrator.c:360]: All this does is check final state is never left. I could remove this or keep it for similarity with prompter. Fine either way.

File: daemon/migrator.c
Comment (Tomin1) [inline daemon/migrator.c:555]: I'm not sure if we need to do this or not. It's also possible that if we can remove all instances of invoker and sailjail from Exec lines, then the existing check (marked with DEAD_CODE) would work as it is.

**Review by Tomin1 — COMMENTED** [review]
We should not merge this before we have replaced uses of sailjail-homescreen-plugin with sailjaild. Hence the draft status.

File: daemon/migrator.c
Comment (spiiroin) [reply daemon/migrator.c:554]: IMO something like: the command line needs to be checked, but that would be something that the party that is going to handle actual execution needs to handle. Here in daemon side we allow/deny running applications, and execution details can and are expected to change without affecting allowed state -> we should ignore binary path in migration.

File: daemon/migrator.c
Comment (spiiroin) [inline daemon/migrator.c:575]: This does ownership transfer of dynamically allocated string,  allocated in migrator_build_queue() and eventually freed in migrator_removal_dequeue_all() / migrator_removal_clear_queue(), right?

Which causes mentail disconnect/confusion in my head to have functions in between those two ends (migrator_process_file and migrator_removal_enqueue) to claim the data to be of 'const char *' type.

File: daemon/migrator.c
Comment (spiiroin) [inline daemon/migrator.c:778]: nit / matter of opinion: I'd prefer to see error handling first. Basically so that when one is reading code it is immediately obvious that error case is handled and it can then be forgotten about vs. doing the other way around and one ends up with stack of error handlers that are far away from where it happened. In this case the situation is short and shallow so it does not really matter...

File: daemon/migrator.c
Comment (spiiroin) [inline daemon/migrator.c:785]: This looks like accident waiting to happen. Maybe something like ...

    size_t n = strlen(HOMESCREEN_DATA_PATH);
    return !strncmp(dir_path, HOMESCREEN_DATA_PATH,n) && dir_path[n] != 0;

Comment (spiiroin) [reply daemon/migrator.c:785]: Or, since removal/ignoring of HOMESCREEN_DATA_PATH prefix seems to have more use below too, one might have a helper function for that purpose and use it instead of making assumptions about the prefix being there?

File: daemon/migrator.c
Comment (spiiroin) [inline daemon/migrator.c:792]: The profile_from_approval_path() function name confuses me. I would assume it returns path to a (profile) file, but clearly that is not the case.

File: daemon/migrator.c
Comment (spiiroin) [inline daemon/migrator.c:802]: nit: gint64 is likely to be wider than uid_t -> having a (uid_t) cast would signal that it is intentional.

**Review by spiiroin — CHANGES_REQUESTED** [review]
Mostly LGTM, but I'd like to see those HOMESCREEN_DATA_PATH assumptions go away and fixing that potential g_source_remove(0) call.

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c:575]: There is `g_strdup()` in `migrator_removal_enqueue()` so no, it doesn't take the ownership. However, I see now that I don't free this value I give to `migrator_process_file()` at all.

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c:792]: What it should do is return `/usr/share/applications/jolla-gallery.desktop` when given `/var/lib/sailjail-homescreen/100000/usr/share/applications/jolla-gallery.desktop/X-Sailjail`. The intention is to extract that path to a real application profile from the approval file path.

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c:404]: Yeah, I guess my previous code was relying on luck that it has been scheduled already. I'll add a later call.

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c:470]: I'll add check that the name is X-Sailjail or Sailjail.

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c]: Added that helper function and rewrote the rest. It should be better now.

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c]: Added that while I rewrote the function around that new `without_leading_data_path()` helper.

**Comment by Tomin1** [comment]
Rebased on master and found some things to improve on my own. This now skips MIGRATED state if there is nothing to migrate. Thinking behind that is that if there is nothing to change, there is no migrated signal and thus settings doesn't need to save anything. (It seems that there is a redundant save action on start anyway, I'm not sure where that comes from but at least it's not from migrator.)

File: daemon/migrator.c
Comment (spiiroin) [reply daemon/migrator.c:575]: Rebase whacked the diff a bit but I think there is that free now (and having it fixes my const confusion too).

File: daemon/migrator.c
Comment (spiiroin) [reply daemon/migrator.c]: ok, it is just a case of a bit peculiar starting path. Might be nice to have a comment to that effect in code just to make reading a bit lighter task.

File: daemon/migrator.c
Comment (spiiroin) [reply daemon/migrator.c:409]: Looks like later call is there.

File: daemon/migrator.c
Comment (spiiroin) [inline daemon/migrator.c:786]: All other case of \#define FOO_PATH omit trailing "/" -> this *looks* like it might stumble like:

- /var/lib/sailjail-homescreenfoo/bar -> foo/bar

But as it is \#define HOMESCREEN_DATA_PATH "/var/lib/sailjail-homescreen/" and that slash is included in strlen result -> that trap is avoided, but to me it looks like it susceptible to skip-too-much problem:

- /var/lib/sailjail-homescreen/foo/bar -> foo/bar

Dropping the trailing slash from macro and inserting it in here might be clearer?

    #define HOMESCREEN_DATA_PATH "/var/lib/sailjail-homescreen"

    if( path && g_str_has_prefix(path, HOMESCREEN_DATA_PATH "/") )
        return path + strlen(HOMESCREEN_DATA_PATH);

(with only slight case of "strncmp way would avoid effort and memory use doubling" tingling somewhere behind ear)

**Review by spiiroin — APPROVED** [review]
Check that without_leading_data_path() comment - is it a real issue or not.

But apart from that I think all my concerns were addressed -> LGTM.

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c]: There is a comment there already so maybe you should check if it's good enough.

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c]: It was correct already AFAICT. I added a comment what it is doing which should help understanding. I also swapped it to use  strncmp instead and made that "/" although it's needed everywhere that this define is used at but at least it's more consistent.

**Comment by Tomin1** [comment]
Rebased. Sorry that it messed up the comparison.

**Comment by Tomin1** [comment]
Since we are getting closer to using sailjaild for sailjail, I would like to get a review from @izh1979, @okodron or @dseight. Please review.

**Review by dseight — APPROVED** [review]

File: daemon/migrator.c
Comment (Tomin1) [reply daemon/migrator.c]: Yeap. Well spotted, thanks. I improved it. It should function the same still.

Also rebased on top of master.

**Review by okodron — APPROVED** [review]
LGTM

**Comment by izh1979** [comment]
LGTM.

**Pull request merged**
